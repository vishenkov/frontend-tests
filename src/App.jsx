import React from 'react';
import {
  Tab,
  Tabs,
  TabList,
  TabPanel,
} from 'react-tabs';

import 'react-tabs/style/react-tabs.css';

export default () => (
  <div className="container">
    <Tabs>
      <TabList>
        <Tab><span id="tab1">Jest</span></Tab>
        <Tab><span id="tab2">Enzyme</span></Tab>
      </TabList>

      <TabPanel>
        <p className="content content1">Jest is a delightful JavaScript Testing Framework with a focus on simplicity.</p>
      </TabPanel>
      <TabPanel>
        <p>
          Enzyme is a JavaScript Testing utility for React that makes
          it easier to test your React Components output.
          You can also manipulate, traverse, and in some ways simulate runtime given the output.
        </p>
        <p>
          Enzyme API is meant to be intuitive and flexible by mimicking
          jQuery API for DOM manipulation and traversal.
        </p>
      </TabPanel>
    </Tabs>
  </div>
);
