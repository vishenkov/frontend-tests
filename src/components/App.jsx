import React from 'react';
import {
  Tab,
  Tabs,
  TabList,
  TabPanel,
} from 'react-tabs';

import 'react-tabs/style/react-tabs.css';
import withStateLogger from '../hocs/withStateLogger';

class App extends React.Component {
  state = {
    hello: 0
  };

  handleClick = () => {
    this.setState({
      hello: this.state.hello + 1
    });
  }

  render() {
    return (
    <div className="container">
    {this.state.hello}
      <Tabs>
        <TabList>
          <Tab data-test="section-tab">Jest</Tab>
          <Tab data-test="section-tab" onClick={this.handleClick}>Enzyme</Tab>
        </TabList>

        <TabPanel>
          <div data-test="section-content">
            <p className="content content1">Jest is a delightful JavaScript Testing Framework with a focus on simplicity.</p>
          </div>
        </TabPanel>
        <TabPanel>
          <div data-test="section-content">
            <p>
              Enzyme is a JavaScript Testing utility for React that makes
              it easier to test your React Components output.
              You can also manipulate, traverse, and in some ways simulate runtime given the output.
            </p>
            <p>
              Enzyme API is meant to be intuitive and flexible by mimicking
              jQuery API for DOM manipulation and traversal.
            </p>
          </div>
        </TabPanel>
      </Tabs>
    </div>
  );
  }
}

export default withStateLogger(App);
