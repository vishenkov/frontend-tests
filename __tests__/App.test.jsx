import _ from 'lodash';
import React from 'react';
import { mount } from 'enzyme';

import App from '../src/components/App';

const AppPage = wrapper => ({
  wrapper,
  tabs() {
    return wrapper.find('li[data-test="section-tab"]');
  },

  tabContent() {
    return wrapper.find('div[data-test="section-content"]');
  },
  firstTab() {
    return this.tabs().at(0);
  },
  secondTab() {
    return this.tabs().at(1);
  },
  removeButtons() {
    return wrapper.find('button[data-test="section-delete-button"]');
  },
  addButton() {
    return wrapper.find('button[data-test="section-tab-add"]').at(0);
  },
  reload() {
    wrapper.unmount();
    wrapper.mount();
  },
});


describe('Matchers suite', () => {
  it('Should switch to second tab', () => {
    const app = AppPage(mount(<App />));
    app.secondTab().simulate('click');

    expect(app.firstTab()).not.toHaveClassName('active');
    expect(app.secondTab()).toHaveClassName('active');
  });


  it('should remove tab on click', () => {
    const app = AppPage(mount(<App />));

    const tabs = app.tabs();
    expect(app.wrapper).toContainMatchingElements(tabs.length, 'li[data-test="section-tab"]');
    const removeButton = app.removeButtons().at(0);

    removeButton.simulate('click');

    expect(app.wrapper).toContainMatchingElements(tabs.length - 1, 'li[data-test="section-tab"]');
  });


  it('should add tab on click', () => {
    const app = AppPage(mount(<App />));

    const tabs = app.tabs();
    expect(app.wrapper).toContainMatchingElements(tabs.length, 'li[data-test="section-tab"]');

    app.addButton().simulate('click');

    expect(app.wrapper).toContainMatchingElements(tabs.length + 1, 'li[data-test="section-tab"]');
  });
});


describe('Storage tests', () => {
  it('Should not fail on empty storage state', () => {
    const storage = {
      get: _.constant(undefined),
      set: _.noop,
    };
    const app = AppPage(mount(<App storage={storage} />));

    expect(app.firstTab()).toHaveClassName('active');
  });


  it('should get active tab from storage', () => {
    const storage = {
      get: _.constant(1),
      set: _.noop,
    };

    const app = AppPage(mount(<App storage={storage} />));
    expect(app.secondTab()).toHaveClassName('active');
  });


  it('should save state on reload', () => {
    const storage = {
      state: {},
      get(key) { return this.state[key]; },
      set(key, value) { this.state[key] = value; },
    };

    const app = AppPage(mount(<App storage={storage} />));
    app.secondTab().simulate('click');
    app.reload();
    expect(app.secondTab()).toHaveClassName('active');
  });
});
