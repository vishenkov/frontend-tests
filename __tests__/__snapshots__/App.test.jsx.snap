// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot suite should change tab on click 1`] = `
<div
  data-test="section-content"
>
  <p>
    Enzyme is a JavaScript Testing utility for React that makes it easier to test your React Components output. You can also manipulate, traverse, and in some ways simulate runtime given the output.
  </p>
  <p>
    Enzyme API is meant to be intuitive and flexible by mimicking jQuery API for DOM manipulation and traversal.
  </p>
</div>
`;
